### PR TITLE
Core/Misc: DB size optimization, don't store default values to database

### PR DESCRIPTION
**Pull request by ariel-**
**Changes proposed**: 
- Move usually unused columns to separate tables.
- Don't store default values in database.
- Split up and unblob several characters and item_instance fields, most of them were always default values.
- This will save data for cases where the values are distinct from default, producing a noticeable decrease in size of the characters DB

**Target branch(es)**: 335

**Tests performed**: (Does it build? Tested in-game?) Build correct with MSVC2015upd1, tried to break it as much as I could (deleting and restoring character, then perm delete, etc.)

**Tests by Aokromes**:

```
Original Database ~50k characters:
1556M Feb 19 07:02 characters.sql

Optimized Database:
1241M Feb 19 17:10 characters_optimized.sql
```

**Tests by Treeston**:

```
Original Database test 2 97 characters:
before 3.2mb

Optimized Database test 2:
615.4kb
```

**Tests by xjose93**:

```
Original DB:
4680.646 MB
Optimized DB:
4065.738 MB

Character count: 128533
```

**Known issues and TODO list**: None as I know.

### CODE REVIEW

**Comment by DDuarte** [comment]
Judging by Aokromes' reports the optimizations in the DB size are very clear (the size needed in a beefy database is reduced a lot) however it's not clear if the optimizations have any impact the query performance.

File: sql/updates/characters/2016_02_19_00_characters.sql [hunk 1]
```diff
@@ -0,0 +1,274 @@
+--
```
Comment (DDuarte) [inline sql/updates/characters/2016_02_19_00_characters.sql]: Missing updates on `sql/base/characters_database.sql`.

**Comment by ariel-** [comment]
The performance shouldn't matter too much, as most of queries are executed async, maybe for the client is some extra delay when logging in, but shouldn't be much of a difference core-wise.

**Comment by Aokromes** [comment]
Also, shauren said: Shauren> probably a bit slower queries but these are joins on indexs so its not THAT bad

**Comment by jackpoz** [comment]
please profile your changes and compare them with the previous behavior.

**Comment by ariel-** [comment]
@jackpoz: How to do that? I'm using windows 64 bit and MSVC 2015 upd 1

**Comment by jackpoz** [comment]
check MySQL documentation and see if they provide any tool to profile query performance

**Comment by xjose93** [comment]
From 4680,646mb to 4065,738mb.

**Comment by AriDEV** [comment]
´´´
From: 1448.960mb
To: 1159.168mb
´´´

**Comment by ariel-** [comment]
Profiling data (thanks @AriDEV)

```
Before optimization: 54.013 ms
After optimization: 60.84 ms
```

Of course the results of only one run with each core can't be taken into account, but here it is for reference.

On each core (restored DB, applied update manually and did the following steps);
 - Login account
 - Login with character 1
 - Logout character
 - Delete character 2 (worldserver set to "unlink" delete)
 - Logout account
 - In server console I do .character deleted restore 2
 - Login account again
 - Delete character 3
 - Logout account, now in server console:
 - .character deleted list
 - .character deleted delete 3
 - .character erase 2
 - .saveall

File: src/server/game/Entities/Player/Player.h [hunk 2]
```diff
@@ -2097,8 +2092,9 @@ class Player : public Unit, public GridObject<Player>
         uint8 m_forced_speed_changes[MAX_MOVE_TYPE];
 
         bool HasAtLoginFlag(AtLoginFlags f) const { return (m_atLoginFlags & f) != 0; }
-        void SetAtLoginFlag(AtLoginFlags f) { m_atLoginFlags |= f; }
```
Comment (Vincent-Michael) [inline src/server/game/Entities/Player/Player.h]: revert this rename pls

**Comment by Aokromes** [comment]
@ariel- can you update this? :)

**Comment by jackpoz** [comment]
this PR has come down to the question: do we consider database space reduction at the cost of query performance (and database complexity with multiple joins instead of denormalized database) an optimization ?

**Comment by jackpoz** [comment]
it is db complexity if I have to write a SELECT with 10 JOINs instead of a simple SELECT with 0 JOINs to get the same data

**Comment by Aokromes** [comment]
What i mean is, the pr is "DB size optimization, don't store default values to database" not "DB complexity reduction" if you have a magic way to reduce db size without increase complexity it's welcomed. Anyway it also reduces complexity of some tables. (killing blobs)

**Comment by jackpoz** [comment]
as written at https://github.com/TrinityCore/TrinityCore/pull/16650#issuecomment-226763214 , it's not a matter of finding a way to reduce db size without increasing complexity, it's a matter to decide if we want to decrease the db size at all at cost of the 2 drawbacks described

**Comment by Aokromes** [comment]
as i wrote on https://github.com/TrinityCore/TrinityCore/pull/16650#issuecomment-226764480 it's not the 1st time we merge such things, quest_template_addon, split character fields, etc, if those where merged, i don't see why this cannot be merged.

**Comment by jackpoz** [comment]
where did you read it cannot be merged ?

**Comment by Aokromes** [comment]
@ariel- irc :P

**Comment by malcrom** [comment]
Default assumes all have same value and then we can't get stuff to work because its the same for all creatures. Like right now having some of an entry fly while others walk.

**Pull request closed without merging**